akka-http-aggregator/src/main/scala/com/pagerduty/akka/http/aggregator/aggregator/GenericAggregator.scala

```diff
@@ -0,0 +1,133 @@
+package com.pagerduty.akka.http.aggregator.aggregator
+
+import akka.http.scaladsl.model.{HttpRequest, HttpResponse}
+import akka.http.scaladsl.unmarshalling.Unmarshal
+import akka.stream.Materializer
+import com.pagerduty.akka.http.aggregator.AggregatorUpstream
+import com.pagerduty.akka.http.headerauthentication.model.HeaderAuthConfig
+import com.pagerduty.akka.http.proxy.HttpProxy
+
+import scala.concurrent.{ExecutionContext, Future}
+
+trait GenericAggregator[RequestKey, AccumulatedState, AddressingConfig]
+    extends Aggregator[AddressingConfig] {
+  type RequestMap =
+    Map[RequestKey, (AggregatorUpstream[AddressingConfig], HttpRequest)]
+  type ResponseMap = Map[RequestKey, (HttpResponse, String)]
+  type AccumulatedStateAndRequests = (AccumulatedState, RequestMap)
+  type HandlerResult = Either[HttpResponse, AccumulatedStateAndRequests]
+  type ResponseHandler = (AccumulatedState, ResponseMap) => HandlerResult
+
+  // implement these methods
+  def handleIncomingRequest(
+      authConfig: HeaderAuthConfig
+  )(incomingRequest: HttpRequest, authData: authConfig.AuthData): HandlerResult
+
+  def intermediateResponseHandlers: Seq[ResponseHandler]
+
+  def buildOutgoingResponse(accumulatedState: AccumulatedState,
+                            upstreamResponses: ResponseMap): HttpResponse
+
+  // implementation
+  def execute(authConfig: HeaderAuthConfig)(authedRequest: HttpRequest,
```

Comment:
This code copied out of `AggregatorController` and updated with the short-circuit pattern.